File: process_image.py
import cv2
import numpy as np
import argparse
import sys
import os

def apply_clahe(img):
    if len(img.shape) == 2:
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8,8))
        return clahe.apply(img)
    elif len(img.shape) == 3:
        lab = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)
        l, a, b = cv2.split(lab)
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8,8))
        cl = clahe.apply(l)
        merged = cv2.merge((cl, a, b))
        return cv2.cvtColor(merged, cv2.COLOR_LAB2BGR)
    return img

def to_black_white(img):
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    _, bw = cv2.threshold(gray, 127, 255, cv2.THRESH_BINARY)
    return bw

def to_grayscale(img):
    return cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

def enhance_channel(img, channel):
    img = img.copy()
    idx = {'r':2, 'g':1, 'b':0}[channel]
    img[:,:,idx] = np.clip(img[:,:,idx] * 1.5, 0, 255).astype(np.uint8)
    return img

def resize_image(img, width, height):
    return cv2.resize(img, (width, height))

def crop_image(img, x, y, w, h):
    return img[y:y+h, x:x+w]

def zoom_image(img, factor):
    h, w = img.shape[:2]
    nh, nw = int(h*factor), int(w*factor)
    return cv2.resize(img, (nw, nh))

def rotate_image(img, angle):
    h, w = img.shape[:2]
    M = cv2.getRotationMatrix2D((w/2, h/2), angle, 1)
    return cv2.warpAffine(img, M, (w, h))

def main():
    parser = argparse.ArgumentParser(description='Process an image.')
    parser.add_argument('--input', required=True, help='Input image path')
    parser.add_argument('--output', required=True, help='Output image path')
    parser.add_argument('--operation', required=True, choices=['clahe', 'bw', 'grayscale', 'enhance_r', 'enhance_g', 'enhance_b', 'resize', 'crop', 'zoom', 'rotate'])
    parser.add_argument('--width', type=int)
    parser.add_argument('--height', type=int)
    parser.add_argument('--x', type=int)
    parser.add_argument('--y', type=int)
    parser.add_argument('--w', type=int)
    parser.add_argument('--h', type=int)
    parser.add_argument('--factor', type=float)
    parser.add_argument('--angle', type=float)
    args = parser.parse_args()

    if not os.path.exists(args.input):
        print('Input file does not exist', file=sys.stderr)
        sys.exit(1)
    img = cv2.imread(args.input)
    if img is None:
        print('Failed to load image', file=sys.stderr)
        sys.exit(1)

    if args.operation == 'clahe':
        out = apply_clahe(img)
    elif args.operation == 'bw':
        out = to_black_white(img)
    elif args.operation == 'grayscale':
        out = to_grayscale(img)
    elif args.operation == 'enhance_r':
        out = enhance_channel(img, 'r')
    elif args.operation == 'enhance_g':
        out = enhance_channel(img, 'g')
    elif args.operation == 'enhance_b':
        out = enhance_channel(img, 'b')
    elif args.operation == 'resize':
        if args.width is None or args.height is None:
            print('Width and height required for resize', file=sys.stderr)
            sys.exit(1)
        out = resize_image(img, args.width, args.height)
    elif args.operation == 'crop':
        if None in (args.x, args.y, args.w, args.h):
            print('x, y, w, h required for crop', file=sys.stderr)
            sys.exit(1)
        out = crop_image(img, args.x, args.y, args.w, args.h)
    elif args.operation == 'zoom':
        if args.factor is None:
            print('Factor required for zoom', file=sys.stderr)
            sys.exit(1)
        out = zoom_image(img, args.factor)
    elif args.operation == 'rotate':
        if args.angle is None:
            print('Angle required for rotate', file=sys.stderr)
            sys.exit(1)
        out = rotate_image(img, args.angle)
    else:
        print('Unknown operation', file=sys.stderr)
        sys.exit(1)

    # Save output
    if len(out.shape) == 2:
        cv2.imwrite(args.output, out)
    else:
        cv2.imwrite(args.output, cv2.cvtColor(out, cv2.COLOR_BGR2RGB) if out.shape[2] == 3 else out)

if __name__ == '__main__':
    main() 

File: process_image.php
<?php
require 'backend/db.php';
session_start();

// Get image info from DB using image_id
$image_id = isset($_GET['image_id']) ? $_GET['image_id'] : (isset($_POST['image_id']) ? $_POST['image_id'] : '');
$image = null;
if ($image_id) {
    $stmt = $conn->prepare("SELECT image_path, description FROM Images WHERE image_id = ?");
    $stmt->bind_param("s", $image_id);
    $stmt->execute();
    $result = $stmt->get_result();
    $image = $result->fetch_assoc();
    $stmt->close();
}

$success = '';
$error = '';

if ($_SERVER['REQUEST_METHOD'] === 'POST' && $image) {
    $input_path = $image['image_path'];
    $operation = '';
    $args = [];
    // Only allow one operation per submission for simplicity
    if (isset($_POST['clahe'])) {
        $operation = 'clahe';
    } elseif (isset($_POST['bw'])) {
        $operation = 'bw';
    } elseif (isset($_POST['grayscale'])) {
        $operation = 'grayscale';
    } elseif (!empty($_POST['enhance_r'])) {
        $operation = 'enhance_r';
    } elseif (!empty($_POST['enhance_g'])) {
        $operation = 'enhance_g';
    } elseif (!empty($_POST['enhance_b'])) {
        $operation = 'enhance_b';
    } elseif (!empty($_POST['resize_w']) && !empty($_POST['resize_h'])) {
        $operation = 'resize';
        $args['width'] = (int)$_POST['resize_w'];
        $args['height'] = (int)$_POST['resize_h'];
    } elseif (!empty($_POST['crop_x']) && !empty($_POST['crop_y']) && !empty($_POST['crop_w']) && !empty($_POST['crop_h'])) {
        $operation = 'crop';
        $args['x'] = (int)$_POST['crop_x'];
        $args['y'] = (int)$_POST['crop_y'];
        $args['w'] = (int)$_POST['crop_w'];
        $args['h'] = (int)$_POST['crop_h'];
    } elseif (!empty($_POST['zoom'])) {
        $operation = 'zoom';
        $args['factor'] = (float)$_POST['zoom'];
    } elseif (!empty($_POST['rotate'])) {
        $operation = 'rotate';
        $args['angle'] = (float)$_POST['rotate'];
    }
    if ($operation) {
        $ext = pathinfo($input_path, PATHINFO_EXTENSION);
        $processed_name = 'proc_' . uniqid() . '.' . $ext;
        $output_path = 'images/processed/' . $processed_name;
        $cmd = escapeshellcmd("python3 process_image.py --input " . escapeshellarg($input_path) . " --output " . escapeshellarg($output_path) . " --operation $operation");
        // Add extra args
        foreach ($args as $k => $v) {
            $cmd .= " --$k " . escapeshellarg($v);
        }
        // For enhance_r/g/b, pass channel
        if (in_array($operation, ['enhance_r','enhance_g','enhance_b'])) {
            // No extra arg needed, operation name is enough
        }
        exec($cmd, $output, $ret);
        if ($ret === 0 && file_exists($output_path)) {
            // Insert into ProcessedImages
            $doctor_id = isset($_SESSION['user_id']) ? $_SESSION['user_id'] : 1;
            $processing_type = $operation;
            $now = date('Y-m-d H:i:s');
            $stmt = $conn->prepare("INSERT INTO ProcessedImages (image_id, processing_type, processed_path, processed_by, processed_at) VALUES (?, ?, ?, ?, ?)");
            $stmt->bind_param("issss", $image_id, $processing_type, $output_path, $doctor_id, $now);
            if ($stmt->execute()) {
                $success = "Image processed and saved! <a href='images.html'>Back to Images</a>";
            } else {
                $error = "Failed to save processed image to database.";
            }
            $stmt->close();
        } else {
            $error = "Image processing failed.";
        }
    } else {
        $error = "Please select an operation and provide required parameters.";
    }
}
$conn->close();
?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Process Image</title>
    <link rel="stylesheet" href="css/style.css">
    <style>
        .process-form {
            max-width: 500px;
            margin: 30px auto;
            background: #fff;
            border-radius: 10px;
            box-shadow: 0 4px 16px rgba(0,0,0,0.08);
            padding: 30px;
        }
        .process-form h2 {
            margin-bottom: 20px;
        }
        .process-form label {
            font-weight: 600;
            display: block;
            margin-top: 15px;
        }
        .process-form input[type="number"],
        .process-form input[type="text"] {
            width: 100%;
            padding: 8px;
            margin-top: 5px;
            border-radius: 5px;
            border: 1px solid #ccc;
        }
        .process-form .form-row {
            display: flex;
            gap: 10px;
        }
        .process-form .form-row > div {
            flex: 1;
        }
        .process-form button {
            margin-top: 20px;
        }
        .image-preview {
            text-align: center;
            margin-bottom: 20px;
        }
        .image-preview img {
            max-width: 100%;
            max-height: 300px;
            border-radius: 8px;
            box-shadow: 0 2px 8px rgba(0,0,0,0.07);
        }
    </style>
</head>
<body>
    <div class="main-content">
        <div class="process-form">
            <h2>Process Image</h2>
            <?php if ($success): ?>
                <div class="success"> <?php echo $success; ?> </div>
            <?php elseif ($error): ?>
                <div class="error"> <?php echo $error; ?> </div>
            <?php endif; ?>
            <?php if ($image): ?>
                <div class="image-preview">
                    <img src="<?php echo htmlspecialchars($image['image_path']); ?>" alt="Image to process">
                    <p><?php echo htmlspecialchars($image['description']); ?></p>
                </div>
                <form method="post" enctype="multipart/form-data">
                    <input type="hidden" name="image_id" value="<?php echo htmlspecialchars($image_id); ?>">
                    <label><input type="checkbox" name="clahe"> CLAHE (Contrast Enhancement)</label>
                    <label><input type="checkbox" name="bw"> Black/White</label>
                    <label><input type="checkbox" name="grayscale"> Grayscale</label>
                    <label>Enhance R/G/B:</label>
                    <div class="form-row">
                        <div><input type="number" name="enhance_r" placeholder="Red %" min="0" max="200"></div>
                        <div><input type="number" name="enhance_g" placeholder="Green %" min="0" max="200"></div>
                        <div><input type="number" name="enhance_b" placeholder="Blue %" min="0" max="200"></div>
                    </div>
                    <label>Resize:</label>
                    <div class="form-row">
                        <div><input type="number" name="resize_w" placeholder="Width (px)"></div>
                        <div><input type="number" name="resize_h" placeholder="Height (px)"></div>
                    </div>
                    <label>Crop:</label>
                    <div class="form-row">
                        <div><input type="number" name="crop_x" placeholder="X"></div>
                        <div><input type="number" name="crop_y" placeholder="Y"></div>
                        <div><input type="number" name="crop_w" placeholder="Width"></div>
                        <div><input type="number" name="crop_h" placeholder="Height"></div>
                    </div>
                    <label>Zoom (scale): <input type="number" name="zoom" placeholder="e.g. 1.2" step="0.01" min="0.1" max="5"></label>
                    <label>Rotate (angle): <input type="number" name="rotate" placeholder="Degrees" min="-360" max="360"></label>
                    <button type="submit" class="process-btn">Process & Save</button>
                </form>
            <?php else: ?>
                <p>Image not found.</p>
            <?php endif; ?>
        </div>
    </div>
</body>
</html> 